lib/location.py

```diff
@@ -90,106 +90,113 @@ def geolocator(location):
 
 
 def ipcachewrite(ip_addr, location):
-    cached = os.path.join(IP2LCACHE, ip_addr)
+    """ Write a retrieved ip+location into cache
+        Can stress some filesystems after long term use, see
+        https://stackoverflow.com/questions/466521/how-many-files-can-i-put-in-a-directory
+    """
+    cachefile = os.path.join(IP2LCACHE, ip_addr)
     if not os.path.exists(IP2LCACHE):
         os.makedirs(IP2LCACHE)
-    with open(cached, 'w') as file:
-        file.write(location[0] + ';' + location[1])
+
+    with open(cachefile, 'w') as file:
+        file.write(location[0] + ';' + location[1] + ';' + location[2])
+        # like ip2location format, but reversed
 
 def ipcache(ip_addr):
-    cached = os.path.join(IP2LCACHE, ip_addr)
+    """ Retrieve a location from cache by ip addr
+        Returns a triple of (CITY, REGION, COUNTRY) or None
+    """
+    cachefile = os.path.join(IP2LCACHE, ip_addr)
     if not os.path.exists(IP2LCACHE):
         os.makedirs(IP2LCACHE)
 
-    location = None
-
-    if os.path.exists(cached):
-        location = open(cached, 'r').read().split(';')
-        if len(location) > 3:
-            return location[3], location[1]
-        elif len(location) > 1:
-            return location[0], location[1]
-        else:
-            return location[0], None
+    if os.path.exists(cachefile):
+        try:
+            city, region, country = open(cachefile, 'r').read().split(';')
```

Comment:
Here I would use this code instead:

```
        if location and ';' in location:
            _, country, region, city = location.split(';')
```

Reply:
I've made this change, now including latlong. See L119 in new file